model/fn-execution/src/main/proto/beam_fn_api.proto

```diff
@@ -375,6 +375,9 @@ message ProcessBundleSplitRequest {
 
     // A set of allowed element indices where the SDK may split. When this is
     // empty, there are no constraints on where to split.
+    // Specifically, the first_residual_element of a split result must be an
+    // allowed split point, and the last_primary_element must immediately
```

Comment:
the last_primary_element shouldn't need to precede the allowed split point as the runner should be able to handle reading N elements from its source + the set of primaries returned from the split if the primary portion needs to be re-executed.